resource/unarchive/unarchive.go

```diff
@@ -0,0 +1,383 @@
+	// the temporary fetch/unarchive location. We check whether the length is 1
+	// because the directory is in the contents.
+	if len(u.destContents) == 1 {
+		return false, nil
+	}
+	return true, nil
+}
+
+// evaluateDuplicates evaluates whether identical files exist in u.Destination
+// and the temporary fetch/unarchive location
+func (u *Unarchive) evaluateDuplicates() error {
+	// determine which directory has fewer items in order to minimize operations
+	dirA := u.destDir.Name()
+	dirB := u.fetchDir.Name()
+	filesA := u.destContents
+	filesB := u.fetchContents
+	if len(u.fetchContents) < len(u.destContents) {
+		dirA = u.fetchDir.Name()
+		dirB = u.destDir.Name()
+		filesA = u.fetchContents
+		filesB = u.destContents
+	}
+
+	// for each item in filesA, determine if it also exists in filesB
+	// compare the checksums for the files - if mismatch, return an error
+	for _, fA := range filesA {
+		for _, fB := range filesB {
+			fileA := strings.TrimPrefix(fA, dirA)
+			fileB := strings.TrimPrefix(fB, dirB)
+
+			faStat, err := os.Stat(fA)
+			if err != nil {
+				return err
+			}
+			fbStat, err := os.Stat(fB)
+			if err != nil {
+				return err
+			}
+
+			if !faStat.IsDir() && !fbStat.IsDir() && fileA == fileB {
+				checkA, err := u.getChecksum(fA)
+				if err != nil {
+					return err
+				}
+
+				checkB, err := u.getChecksum(fB)
+				if err != nil {
+					return err
+				}
+
+				if checkA != checkB {
+					return fmt.Errorf("will not replace, %q exists at %q: checksum mismatch", fileA, u.Destination)
+				}
+
+				break
+			}
+		}
+	}
+
+	return nil
+}
+
+// copyToFinalDest copies the fetched and unarchived files from their temporary
+// directory to the final destination
+func (u *Unarchive) copyToFinalDest() error {
+	// for each item in the fetchDir, mkdir or copy to the final destination
+	for _, file := range u.fetchContents {
+		src, err := os.Open(file)
+		if err != nil {
+			return err
+		}
+		defer src.Close()
+
+		fileName := strings.TrimPrefix(file, u.fetchDir.Name())
+
+		fStat, err := os.Stat(file)
+		if err != nil {
+			return err
+		}
+
+		if fileName != "" {
+			if fStat.IsDir() {
+				err = os.Mkdir(u.destDir.Name()+fileName, fStat.Mode().Perm())
+				if err != nil {
+					if !os.IsNotExist(err) {
+						continue
+					}
+					return err
+				}
+			} else {
+				// get src []byte
+				srcData, err := ioutil.ReadAll(src)
+				if err != nil {
+					return err
+				}
+
+				// get src FileInfo
+				srcInfo, err := src.Stat()
+				if err != nil {
+					return err
+				}
+
+				err = ioutil.WriteFile(u.destDir.Name()+fileName, srcData, srcInfo.Mode().Perm())
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// setFetchLoc sets the location for the fetch destination
+func (u *Unarchive) setFetchLoc() {
+	if u.fetchLoc != "" {
+		return
+	}
+
+	_, file := filepath.Split(u.Source)
+	u.fetchLoc = "/var/run/converge/cache/" + file
```

Comment:
Maybe this should be some sort of var that we can change/set easily, especially if we decide to move it around based on the UID of the user running converge to avoid permission problems.